### PR TITLE
388 drag and drop does not work

### PR DESCRIPTION
**Pull request by vin-ni**
Fixing the ordering of the graphs on `Save Scene`

### CODE REVIEW

**Review by capital-G — CHANGES_REQUESTED** [review]
This does not prevent the problems when one shifts the order and then clicks "please create a new script cell", as then an update from the server comes (via the subscription) which neglects the unsaved changes.

There are 2 ways to account for this

* set the `scriptCellsModified` store variable to true which prevents any kind of synchronization from the server
* or, IMO better, trigger a mutation when switching the order (and lock the editor while waiting for an update for the server in order to prevent corrupted states/"state race conditions" when saving)

The 2nd option accounts for fewer differences between client and server state which can prevent many "merge conflicts" and unexpected events.

**Comment by vin-ni** [comment]
> * or, IMO better, trigger a mutation when switching the order (and lock the editor while waiting for an update for the server in order to prevent corrupted states/"state race conditions" when saving)

Yes, I'd be opting for that as well. Let me propose something

**Comment by capital-G** [comment]
> > * or, IMO better, trigger a mutation when switching the order (and lock the editor while waiting for an update for the server in order to prevent corrupted states/"state race conditions" when saving)
> 
> Yes, I'd be opting for that as well. Let me propose something

Essentially you want a callback on each "update" from the subscription which resets the "unsaved state" flag.
This can be either done via a watcher on the subscription (not sure if this works) or use some urql built-in thing called a `handler`, see https://formidable.com/open-source/urql/docs/api/svelte/#subscriptionstore

**Comment by vin-ni** [comment]
@capital-G I added a simple solution to save the whole node on drag & drop. 

Since the `saveNode` mutation runs rather quick this qould work fine for now.

I am wondering if we can have a custom mutation, which just mutates the orders of the scriptcells. This way I believe what could happen is:

1. User 1: Writing in `scriptcell position 2`
2. User 2: Updates the order and moves it to `position 3`.
3. User 1: Scriptcell moves done one step in real time, but the content stays the interim content.

Would it work like that, or would the subscription pull the whole data of the scriptcell and overwrite the unsaved content of `User 1`?

**Comment by capital-G** [comment]
> @capital-G I added a simple solution to save the whole node on drag & drop.

I don't understand the new emits - they are there to trigger an update as they don't provide any args? But why does a script cell wants an update on its own, it is synced nonetheless through the subscription?
Also who/when triggers the update after a move?

> I am wondering if we can have a custom mutation, which just mutates the orders of the scriptcells. This way I believe what could happen is:

Creating a different route just for updating the order doesn't help in this case as we are bound to the update of the subscription which is the tricky part here as we are distributing the current full state on the server and not diffs (could be possible but it is really hard to implement a good protocol here).

IIRC the update of script cells can be partial (if not we should implement it this way), which means that we should keep track of what has been changed by the user and what has not and only communicate this change/diff to the server which then propagates the current full state to every client.
This allows to change the order in client A and edit the content it client B and get a full state in the end which accounts for the new order and the new content.

But this contradicts the current implementation

https://github.com/Gencaster/gencaster/blob/fa8f079372cca6acad31b44977294b8ca1466cce/caster-editor/src/views/GraphDetailView.vue#L84-L85

which always calculates the new order.

**Comment by capital-G** [comment]
I'll finish this

**Comment by capital-G** [comment]
I think I broke a transition for the editor during merge :/ sorry
Maybe merge this and afterwards add the transition again?
This one is what i am talking about: https://github.com/Gencaster/gencaster/pull/465/files#diff-89fa5c68da41405339f471b0c627de00d039f538de00090e33cd7fe7a1b6b3f7R105

File: caster-editor/src/App.vue [hunk 1]
```diff
@@ -8,7 +8,7 @@ import { useInterfaceStore } from "@/stores/InterfaceStore";
 import { storeToRefs } from "pinia";
 import { ElMessage } from "element-plus";
 
-const {user} = storeToRefs(useInterfaceStore());
+const { user } = storeToRefs(useInterfaceStore());
```
Comment (vin-ni) [inline caster-editor/src/App.vue:11]: I like the new linter. I was looking for that

File: caster-editor/src/components/DialogAddNode.vue [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+const showDialog: Ref<boolean> = ref(true);
+
+const createNodeMutation = useCreateNodeMutation();
+
+const createNode = async () => {
+  let positionX = 0;
+  let positionY = 0;
+  if (vNetworkGraph.value) {
+    const { height, width } = vNetworkGraph.value.getSizes();
+    const pos = vNetworkGraph.value.translateFromDomToSvgCoordinates({
```
Comment (vin-ni) [inline caster-editor/src/components/DialogAddNode.vue:62]: This works? 
I had big problems with that method.

**Review by vin-ni — COMMENTED** [review]
Looks great from my side!

**Pull request merged**